service/gcs/stdio/stdio.go

```diff
@@ -99,6 +99,136 @@ func (s *ConnectionSet) Files() (_ *FileSet, err error) {
 	return fs, nil
 }
 
+// NewPipeRelay returns a new pipe relay wrapping the given connection stdin, stdout, stderr set.
+func (s *ConnectionSet) NewPipeRelay() (_ *PipeRelay, err error) {
+	pr := &PipeRelay{s: s}
+	defer func() {
+		if err != nil {
+			pr.closePipes()
+		}
+	}()
+
+	if s.In != nil {
+		pr.pipes[0], pr.pipes[1], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdin pipe relay")
+		}
+	}
+	if s.Out != nil {
+		pr.pipes[2], pr.pipes[3], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdout pipe relay")
+		}
+	}
+	if s.Err != nil {
+		pr.pipes[4], pr.pipes[5], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stderr pipe relay")
+		}
+	}
+	return pr, nil
+}
+
+// PipeRelay is a relay built to expose a pipe interface
+// for stdin, stdout, stderr ontop of a ConnectionSet.
+type PipeRelay struct {
+	wg sync.WaitGroup
+	s  *ConnectionSet
+	// pipes format is stdin [0 read, 1 write], stdout [2 read, 3 write], stderr [4 read, 5 write].
+	pipes [6]*os.File
```

Comment:
It would be clearer to me to maybe have three sets of pipes (one for stdin, one for stdout, one for stderr). You can represent each pipe as an array or a struct or whatever works, but I just think separating them out makes the code clearer and reduces the risk of errors putting the wrong index somewhere.